### ISSUE TITLE
Dashboard startup fails in Docker

### ISSUE THREAD

**Issue by geoffp** [home-assistant/appdaemon#164]
Hi there!

After following the guide here (http://appdaemon.readthedocs.io/en/latest/DOCKER_TUTORIAL.html), completely fresh, I've got a non-running dashboard. the log output is this:

```
2017-10-21 17:35:27.698343 INFO AppDaemon Version 2.1.10 starting
2017-10-21 17:35:27.698517 INFO Configuration read from: /conf/appdaemon.yaml
2017-10-21 17:35:27.746694 INFO Starting Apps
2017-10-21 17:35:27.774195 INFO Got initial state
2017-10-21 17:35:27.774807 INFO Loading Module: /conf/apps/hello.py
2017-10-21 17:35:27.775939 INFO Loading Object hello_world using class HelloWorld from module hello
2017-10-21 17:35:27.837436 INFO hello_world: Hello from AppDaemon
2017-10-21 17:35:27.838284 INFO hello_world: You are now ready to run Apps!
2017-10-21 17:35:27.838410 INFO App initialization complete
2017-10-21 17:35:27.838600 INFO Starting dashboard
2017-10-21 17:35:27.841010 WARNING ------------------------------------------------------------
2017-10-21 17:35:27.841103 WARNING Unexpected error in dashboard thread
2017-10-21 17:35:27.841170 WARNING ------------------------------------------------------------
2017-10-21 17:35:27.842480 WARNING Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/aiohttp/web_urldispatcher.py", line 401, in __init__
    raise ValueError('Not a directory')
ValueError: Not a directory

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/appdaemon/rundash.py", line 318, in run_dash
    setup_routes(conf.dashboard_obj)
  File "/usr/local/lib/python3.6/site-packages/appdaemon/rundash.py", line 288, in setup_routes
    app.router.add_static('/compiled_javascript', dashboard.compiled_javascript_dir)
  File "/usr/local/lib/python3.6/site-packages/aiohttp/web_urldispatcher.py", line 845, in add_static
    follow_symlinks=follow_symlinks)
  File "/usr/local/lib/python3.6/site-packages/aiohttp/web_urldispatcher.py", line 404, in __init__
    "No directory exists at '{}'".format(directory)) from error
ValueError: No directory exists at '/conf/compiled/javascript'

2017-10-21 17:35:27.842585 WARNING ------------------------------------------------------------
2017-10-21 17:35:27.842655 INFO API is disabled
2017-10-21 17:35:27.848848 INFO Connected to Home Assistant 0.55.2
```

Maybe the `aiohttp` package is missing...?

The server is Ubuntu 16.04.3 LTS, with Docker version 17.05.0-ce, build 89658be. Image is pulled from latest as of today.

**Comment by geoffp** [comment]
On further inspection, it looks like it may be that aiohttp is being handed a path to a nonexistent directory. The indicated line is here: https://github.com/aio-libs/aiohttp/blob/c02e119413c0023c1e6c93c34004799c68fbe92b/aiohttp/web_urldispatcher.py#L401

**Comment by geoffp** [comment]
Yup, my next guess is that the `/compiled_javascript` directory isn't there.

**Comment by geoffp** [comment]
Wrong -- `compiled_javascript` is the name of the route. It's looking to hook that up to `/conf/compiled/javascript`, which indeed doesn't exist.

**Comment by geoffp** [comment]
`/conf` leads out to a host directory. So if I go into that on my host machine and:
```sh
mkdir compiled
mkdir compiled/javascript
mkdir compiled/css
```
It runs! I admit that I juggled some permissions on that before I ran it, but at least it works.

Depending on when the compiled static assets for the web UI get generated (is it on each run of the container?), might it be more appropriate to generate them into a directory *inside* the container? It seems like this stuff doesn't belong in the configuration directory.

**Comment by correctlyu** [comment]
Hello your problem solved, I encountered the same problem

**Comment by acockburn** [comment]
I'l lbe adding some more robust code to check on this at some point - those directories should be created automatically but I suspect there is a permissions issue in Docker preventing that.

**Comment by JoniR** [comment]
Sorry but I didn't get how I should fix this. Where I should create those compiled* directories? Inside of docker image or my server root(which sounds weird)? I tried to create them on `/home/joni/docker/appdaemon/conf` which is my conf directory.
I try to run it like this:
`sudo docker run --rm -it -p 5050:5050 
  -e HA_URL="url"
  -e HA_KEY="key" 
  -e DASH_URL="http://$HOSTNAME:5050" 
  -v /home/joni/docker/appdaemon/conf:/conf 
  acockburn/appdaemon:latest`

**Comment by acockburn** [comment]
They need to be under your config directory -looks like yours is set to /conf - cd to that directory and make the dirs as outlined above.

**Comment by JoniR** [comment]
Thanks - I don't know what I hazzled in the first place but by removing whole conf directory, restart docker and creating compiled* directories, now its working.

**Comment by acockburn** [comment]
Great :)

**Comment by geoffp** [comment]
@acockburn if those compiled files are generated every time the container is run, might we avoid permissions issues entirely by just serving them from inside the container instead of from the config directory?

**Comment by acockburn** [comment]
Unfortunately that won;t work - they are compiled versions of the Dashboard and need to change if changes are made to any of the dashboard config

**Comment by UglyBob79** [comment]
Had the same problem and managed to fix it by creating those folders. But now I can't get my simple Hello.dash to compile, I only get:

2017-12-16 12:24:27.651679 INFO Compiling dashboard 'Hello'
2017-12-16 12:24:27.728138 WARNING Dashboard 'Hello' not found
2017-12-16 12:24:27.728394 WARNING ------------------------------------------------------------
2017-12-16 12:24:27.728526 WARNING Unexpected error during DASH creation
2017-12-16 12:24:27.728672 WARNING ------------------------------------------------------------
2017-12-16 12:24:27.736189 WARNING Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/appdaemon/dashboard.py", line 803, in get_dashboard
    with open(include_path, "r") as include_file:
FileNotFoundError: [Errno 2] No such file or directory: '/conf/compiled/html/default/hello_head.html'

I run AppDaemon in docker and have mapped the config to the host and root owns all the config including the dashboard and the compiled folders... Not sure what else to do.

**Comment by acockburn** [comment]
OK, this is fixed in 3.0 - at some stage the code creating the dirs was moved to after the webserver initialization.
